processor/samplingprocessor/probabilisticsamplerprocessor/probabilisticsampler.go

```diff
@@ -74,49 +70,59 @@ func newTraceProcessor(nextConsumer consumer.TraceConsumer, cfg Config) (compone
+		rspan := rspans.At(i)
+		if rspan.IsNil() {
+			continue
+		}
+		if err := tsp.processTraces(ctx, rspan, sampledTraceData); err != nil {
 			errs = append(errs, err)
 		}
 	}
 	return componenterror.CombineErrors(errs)
 }
 
-func (tsp *tracesamplerprocessor) processTraces(ctx context.Context, td consumerdata.TraceData) error {
+func (tsp *tracesamplerprocessor) processTraces(ctx context.Context, resourceSpans pdata.ResourceSpans, sampledTraceData pdata.Traces) error {
 	scaledSamplingRate := tsp.scaledSamplingRate
 
-	sampledTraceData := consumerdata.TraceData{
-		Node:         td.Node,
-		Resource:     td.Resource,
-		SourceFormat: td.SourceFormat,
-	}
-
-	sampledSpans := make([]*tracepb.Span, 0, len(td.Spans))
-	for _, span := range td.Spans {
-		sp := parseSpanSamplingPriority(span)
-		if sp == doNotSampleSpan {
-			// The OpenTelemetry mentions this as a "hint" we take a stronger
-			// approach and do not sample the span since some may use it to
-			// remove specific spans from traces.
+	sampledTraceData.ResourceSpans().Resize(sampledTraceData.ResourceSpans().Len() + 1)
+	rs := sampledTraceData.ResourceSpans().At(sampledTraceData.ResourceSpans().Len() - 1)
+	rs.Resource().InitEmpty()
+	resourceSpans.Resource().CopyTo(rs.Resource())
+	rs.InstrumentationLibrarySpans().Resize(1)
+	spns := rs.InstrumentationLibrarySpans().At(0).Spans()
+
+	ilss := resourceSpans.InstrumentationLibrarySpans()
+	for j := 0; j < ilss.Len(); j++ {
+		ils := ilss.At(j)
+		if ils.IsNil() {
 			continue
 		}
+		for k := 0; k < ils.Spans().Len(); k++ {
+			span := ils.Spans().At(k)
+			sp := parseSpanSamplingPriority(span)
+			if sp == doNotSampleSpan {
+				// The OpenTelemetry mentions this as a "hint" we take a stronger
+				// approach and do not sample the span since some may use it to
+				// remove specific spans from traces.
+				continue
+			}
 
-		// If one assumes random trace ids hashing may seems avoidable, however, traces can be coming from sources
-		// with various different criteria to generate trace id and perhaps were already sampled without hashing.
-		// Hashing here prevents bias due to such systems.
-		sampled := sp == mustSampleSpan ||
-			hash(span.TraceId, tsp.hashSeed)&bitMaskHashBuckets < scaledSamplingRate
+			// If one assumes random trace ids hashing may seems avoidable, however, traces can be coming from sources
+			// with various different criteria to generate trace id and perhaps were already sampled without hashing.
+			// Hashing here prevents bias due to such systems.
+			sampled := sp == mustSampleSpan ||
+				hash(span.TraceID(), tsp.hashSeed)&bitMaskHashBuckets < scaledSamplingRate
 
-		if sampled {
-			sampledSpans = append(sampledSpans, span)
+			if sampled {
+				spns.Append(&span)
+			}
 		}
 	}
-
-	sampledTraceData.Spans = sampledSpans
-
-	return tsp.nextConsumer.ConsumeTraces(ctx, internaldata.OCToTraceData(sampledTraceData))
+	return tsp.nextConsumer.ConsumeTraces(ctx, sampledTraceData)
```

Comment:
This call should be moved out of the loop in `func (tsp *tracesamplerprocessor) ConsumeTraces`. If I understand this logic correctly it will be double publishing sampled traces the following way:
Original `pdata.Traces{origResourceSpans1, origResourceSpans2, origResourceSpans3}` will cause three pushes to the nextConsumer:
```
pdata.Traces{sampledResourceSpans1}
pdata.Traces{sampledResourceSpans1, sampledResourceSpans2}
pdata.Traces{sampledResourceSpans1, sampledResourceSpans2, sampledResourceSpans3}
```

Reply:
@dmitryax Yes my bad, i missed it. Let me add a testcase to cover this scenario.